constraint-solver/src/rule_based_optimizer/rules.rs

```diff
@@ -271,6 +276,73 @@ crepe! {
       AlgebraicConstraint(e),
       Solvable(e, var, v);
 
+    ///////////////////////////////// CONSTRAINT SPLITTING //////////////////////////
+
+    // Split the expression into head and tail
+    struct ExpressionSumHeadTail(Expr, Expr, Expr);
```

Comment:
I think the `ExpressionSumHeadTail` should go nito the structural section and then I think we should create a new section for the "Minimal Range" rules.